File: java/var-core/src/test/java/com/oselvar/var/core/DiagnosticsTest.java
package com.oselvar.var.core;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertSame;

import org.junit.jupiter.api.Test;

/**
 * Light coverage of {@link Diagnostics}'s two constructor helpers — this module is a thin
 * supporting piece for {@link Plan}, which is where the bulk of the behavior (and tests) lives.
 */
class DiagnosticsTest {

    private static final Span SPAN = new Span(0, 5, 1, 1, 1, 6);

    @Test
    void ambiguousMatchBuildsAnErrorSeverityDiagnosticWithTheGivenSpan() {
        Diagnostics.Diagnostic d = Diagnostics.ambiguousMatch(SPAN);
        assertEquals(Diagnostics.DiagnosticCode.AMBIGUOUS_MATCH, d.code());
        assertEquals(Diagnostics.Severity.ERROR, d.severity());
        assertSame(SPAN, d.span());
    }

    @Test
    void errorFenceWithoutStepBuildsAnErrorSeverityDiagnosticWithTheGivenSpan() {
        Diagnostics.Diagnostic d = Diagnostics.errorFenceWithoutStep(SPAN);
        assertEquals(Diagnostics.DiagnosticCode.ERROR_FENCE_WITHOUT_STEP, d.code());
        assertEquals(Diagnostics.Severity.ERROR, d.severity());
        assertSame(SPAN, d.span());
    }
}

File: rust/core/tests/diagnostics_test.rs
//! Port of `DiagnosticsTest.java`. Java's `assertSame` (identity) becomes value
//! equality — [`Span`] is a `Copy` value type.

use varar_core::diagnostics::{
    DiagnosticCode, Severity, ambiguous_match, error_fence_without_step,
};
use varar_core::span::Span;

const SPAN: Span = Span {
    start_offset: 0,
    end_offset: 5,
    start_line: 1,
    start_col: 1,
    end_line: 1,
    end_col: 6,
};

#[test]
fn ambiguous_match_builds_an_error_severity_diagnostic_with_the_given_span() {
    let d = ambiguous_match(SPAN);
    assert_eq!(DiagnosticCode::AmbiguousMatch, d.code);
    assert_eq!(Severity::Error, d.severity);
    assert_eq!(SPAN, d.span);
}

#[test]
fn error_fence_without_step_builds_an_error_severity_diagnostic_with_the_given_span() {
    let d = error_fence_without_step(SPAN);
    assert_eq!(DiagnosticCode::ErrorFenceWithoutStep, d.code);
    assert_eq!(Severity::Error, d.severity);
    assert_eq!(SPAN, d.span);
}
